sdk-typed-search-attributes.md

```diff
@@ -0,0 +1,497 @@
+* .NET
+  * Still in alpha, can decide whatever we want
+
+### Server
+
+* ElasticSearch
+  * Multi-value search attribute for non-keyword-list - accepted, but discouraged, and returned as given
+    * TODO(cretz): Question outstanding to server team about the benefits of multi-value non-keyword-list attributes
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+* SQL
+  * Multi-value search attribute for non-keyword-list - only allowed if a single value, always returned single value
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+
+### Problems with current behavior
+
+* Go and Java rely on user to resolve ambiguity between single and multi value search attributes (Go is worse because it has no helpers)
+* TypeScript and Python always disambiguate making all search attributes a list (and Java too in single-value cases) even if list not allowed anymore
+* Setting invalid search attributes in workflow never returns a failure that can be handled by workflow code
+* Users have no way to clarify the difference between keyword and keyword-list
+
+## Goals
+
+Immediate:
+
+* As much as is supported by the language, users need to be able to get and set individually typed search attributes
+* Maintain backwards compatibility but deprecate any approach that doesn't let you identify the search attribute type
+  alongside the use of its value
+
+Non-immediate:
+
+* Provide ways to define predefined search attributes that are typed
+* Also provide ways to define predefined memo values that are typed
+
+Non-goals:
+
+* Better validation of actual search attribute type from server
+* Better way for server to tell SDK the upsert failed
+* Support for user-explicit upsert of multiple typed search attributes in one statement
+* Support for typed registration of search attributes
+* Support for predefined search attributes in any form of query building
+
+## Proposed solution
+
+### General
+
+* All SDK workflow APIs will now have a mutable search attribute map with typed getters and setters
+  * When a mutation on a search attribute occurs:
+    * If the most recent command is an upsert search attribute, the single value is merged
+    * If the most recent command is not an upsert search attribute, make one
+    * Map is mutated
+    * TODO(cretz): How are deletes represented?
+* Upsert will be deprecated
+
+### Go
+
+```go
+// In client
+type SearchAttributes map[string]any
+
+// Yes these already exist today with maps themselves, but clear methods may
+// help discourage people from using the map at all. Can remove though.
```

Comment:
If you're gonna do this, why not just hide the map inside a struct like you have with the in-workflow version? IMO it makes sense for it to not look like a normal map, because you can't just directly set stuff on it.

Reply:
I have to continue to accept a map type in `StartWorkflowOptions.SearchAttributes`. So if I change that type from `map[string]interface{}` to this type, this code still works:

```go
myOptions.SearchAttributes = map[string]interface{}{"foo": "bar"}
```